README.md
# opensphere-plugin-geopackage

GeoPackage support for [OpenSphere](https://github.com/ngageoint/opensphere).

## Getting Started

For ease of development, we have created [opensphere-yarn-workspace](https://github.com/ngageoint/opensphere-yarn-workspace)
which assists in developing across multiple projects.

After you set that up, your directory structure should look like this:
```
opensphere-yarn-workspace/ # your clone of opensphere-yarn-workspace
  workspace/
    opensphere/ # your clone of OpenSphere
    opensphere-plugin-geopackage/ # this project
```

Run `yarn` to install dependencies.

## Things that work

- Tiled imagery in configured application projections (generally EPSG:3857 and EPSG:4326 by default)
- Vector features
- Export of vector features to GeoPackage

## Caveats

Building the web version of OpenSphere still limits the file size to 100MB. The [Electron](https://github.com/ngageoint/opensphere-electron) version supports
large GeoPackage files.

## Tips

After loading tiled imagery with a small coverage area, right-click the layer and select 'Go To'. That will get
you close and you can zoom in a little further to see the imagery.


# Electron

## Building native deps on Windows

 1. Ensure python is installed (latest 2.7 should work; 3 might work but is untested on our end)
 2. Install [Microsoft Build Tools 2013](http://www.microsoft.com/en-us/download/details.aspx?id=40760)
 3. Restart
 4. `cd opensphere` and run `yarn build`
 4. `cd ../opensphere-electron` and run `yarn postinstall` and verify that there are no errors
 5. Run the app or create the installers per instructions in [opensphere-electron](https://github.com/ngageoint/opensphere-electron)
 
## Developing with Electron and GeoPackage
  * In order to develop with Electron and GeoPackage, make sure to have canvas installed as a native dependency on your machine. You can install canvas [here](https://github.com/Automattic/node-canvas#compiling).


CHANGELOG.md
## [1.7.10](https://github.com/ngageoint/opensphere-plugin-geopackage/compare/v1.7.9...v1.7.10) (2021-10-06)


### Bug Fixes

* **tile:** detect tile boundaries close to +/- 180 ([b734fce](https://github.com/ngageoint/opensphere-plugin-geopackage/commit/b734fce173d24dc021cf47ca636e7fc6239d75a8))
* **tile:** fix extent normalization for tile boundaries ([325387b](https://github.com/ngageoint/opensphere-plugin-geopackage/commit/325387b510d2b1c810cbecdccc34167e0956b842))
* **tile:** fix resolutions in main process ([f63e7bb](https://github.com/ngageoint/opensphere-plugin-geopackage/commit/f63e7bb30df417c73783b129e3f515e06e0448db))

## [1.7.9](https://github.com/ngageoint/opensphere-plugin-geopackage/compare/v1.7.8...v1.7.9) (2021-10-04)


### Bug Fixes

* **es:** convert requires to imports ([6507649](https://github.com/ngageoint/opensphere-plugin-geopackage/commit/650764952fe3833f71dbaa6ae2e3fb4de99f14df))

## [1.7.8](https://github.com/ngageoint/opensphere-plugin-geopackage/compare/v1.7.7...v1.7.8) (2021-09-22)


### Bug Fixes

* **import:** sort import statements ([6695306](https://github.com/ngageoint/opensphere-plugin-geopackage/commit/66953065da113f4ffc6ec378a749d8825d4a658a))

## [1.7.7](https://github.com/ngageoint/opensphere-plugin-geopackage/compare/v1.7.6...v1.7.7) (2021-09-16)


### Bug Fixes

* **goog:** fix module paths ([de48c47](https://github.com/ngageoint/opensphere-plugin-geopackage/commit/de48c47693dc249d9689605ae510ff0deb1bd734))

## [1.7.6](https://github.com/ngageoint/opensphere-plugin-geopackage/compare/v1.7.5...v1.7.6) (2021-08-25)


### Bug Fixes

* **export:** support bulk export from multiple sources ([4b7b66c](https://github.com/ngageoint/opensphere-plugin-geopackage/commit/4b7b66cee03ae7828b0bbd84ea007a917b9571b3))

## [1.7.5](https://github.com/ngageoint/opensphere-plugin-geopackage/compare/v1.7.4...v1.7.5) (2021-08-18)


### Bug Fixes

* **import:** provide appropriate types to the compiler ([837d4e3](https://github.com/ngageoint/opensphere-plugin-geopackage/commit/837d4e3726a4c168b758a7e6d741e778baac6cde))

## [1.7.4](https://github.com/ngageoint/opensphere-plugin-geopackage/compare/v1.7.3...v1.7.4) (2021-07-29)


### Bug Fixes

* **tiles:** enables and fixes tile world wrapping ([dd2e6d6](https://github.com/ngageoint/opensphere-plugin-geopackage/commit/dd2e6d637ce58bc15d3976d3be67fd34e3ea6dfb))

## [1.7.3](https://github.com/ngageoint/opensphere-plugin-geopackage/compare/v1.7.2...v1.7.3) (2021-07-26)


### Bug Fixes

* resolve compiler typing errors ([02c7f8c](https://github.com/ngageoint/opensphere-plugin-geopackage/commit/02c7f8c01ef844503f45ecd27f8c2acb352afd02))

## [1.7.2](https://github.com/ngageoint/opensphere-plugin-geopackage/compare/v1.7.1...v1.7.2) (2021-07-22)


### Bug Fixes

* **columns:** filters out empty geometry column ([649b319](https://github.com/ngageoint/opensphere-plugin-geopackage/commit/649b31978890b15323e95166b2b8544b9017696e))

## [1.7.1](https://github.com/ngageoint/opensphere-plugin-geopackage/compare/v1.7.0...v1.7.1) (2021-07-19)


### Bug Fixes

* **tile:** don't destructure default exports ([7579fd3](https://github.com/ngageoint/opensphere-plugin-geopackage/commit/7579fd373e4d20991e1ae406b375ee5b56113af7))

# [1.7.0](https://github.com/ngageoint/opensphere-plugin-geopackage/compare/v1.6.2...v1.7.0) (2021-07-13)


### Features

* **deps:** upgrade better-sqlite to support electron 13 prebuilt binaries ([ab9cb35](https://github.com/ngageoint/opensphere-plugin-geopackage/commit/ab9cb3542310b764b6b4bae154d777d82a00fe8e))

## [1.6.2](https://github.com/ngageoint/opensphere-plugin-geopackage/compare/v1.6.1...v1.6.2) (2021-07-08)


### Bug Fixes

* **columns:** create source columns with the same name used by features ([9be0b94](https://github.com/ngageoint/opensphere-plugin-geopackage/commit/9be0b94992da7f71e4ab0c2e0b13afc183ad39a7))

## [1.6.1](https://github.com/ngageoint/opensphere-plugin-geopackage/compare/v1.6.0...v1.6.1) (2021-07-06)


### Bug Fixes

* **tiles:** only zoom to +4 zoom levels ([b7177c1](https://github.com/ngageoint/opensphere-plugin-geopackage/commit/b7177c1a5a257568552e8b6b5debf78164f307ca))
* **tiles:** use tile scaling to support gpkg tiles at all zooms ([b7097cb](https://github.com/ngageoint/opensphere-plugin-geopackage/commit/b7097cb8cdce2c73cf5107b86319c61b7faccfd8))

# [1.6.0](https://github.com/ngageoint/opensphere-plugin-geopackage/compare/v1.5.3...v1.6.0) (2021-07-02)


### Features

* replace global refs and trigger release ([338cc29](https://github.com/ngageoint/opensphere-plugin-geopackage/commit/338cc29e69392ce5a83b397d3625be6db541002b))

## [1.5.3](https://github.com/ngageoint/opensphere-plugin-geopackage/compare/v1.5.2...v1.5.3) (2021-05-27)


### Bug Fixes


src/worker/gpkg.worker.js
/**
 * Worker to provide GPKG access to avoid blocking the main thread during database interactions
 */
'use strict';

var GeoPackage;


/**
 * This corresponds to plugin.geopackage.MsgType
 * @enum {string}
 */
var MsgType = {
  SUCCESS: 'success',
  ERROR: 'error'
};


/**
 * @type {boolean}
 */
var isNode = false;


/**
 * placeholder for library
 */
var GeoPackage = null;


/**
 * The left boundary to use when normalizing a tile extent. This is used to account for Cesium creating tile bounds
 * very close to +/- 180.
 * @type {number}
 */
const tileLeftBoundary = -180 - 1E-12;


/**
 * The right boundary to use when normalizing a tile extent. This is used to account for Cesium creating tile bounds
 * very close to +/- 180.
 * @type {number}
 */
const tileRightBoundary = 180 + 1E-12;


/**
 * Normalize a tile boundary extent.
 * @param {Array<number>} extent The extent.
 */
const normalizeTileExtent = function(extent) {
  // Make sure we start with numbers in the proper order (left < right).
  let left = Math.min(extent[0], extent[2]);
  let right = Math.max(extent[0], extent[2]);

  if (right - left >= 360) {
    // Whole world, just use +/- 180.
    left = -180;
    right = 180;
  } else if (left < tileLeftBoundary) {
    // Wrapped left, shift right into +/- 180.
    while (left < tileLeftBoundary) {
      left += 360;
      right += 360;
    }
  } else if (right > tileRightBoundary) {
    // Wrapped right, shift left into +/- 180.
    while (right > tileRightBoundary) {
      left -= 360;
      right -= 360;
    }
  }

  extent[0] = left;
  extent[2] = right;
};


/**
 * @return {boolean} Whether or not the OS is Windows
 */
var isWin = function() {
  return isNode ? process.platform === 'win32' :
    /Windows/.test(self.navigator.userAgent);
};


/**
 * @param {Error|string} reason
 * @param {GeoPackageWorkerMessage} originalMsg
 */
var handleError = function(reason, originalMsg) {
  if (reason instanceof Error) {
    reason = reason.toString();
  }

  // don't send anything potentially large back in the error message
  delete originalMsg.data;

  postMessage({type: MsgType.ERROR, reason: reason, message: originalMsg});
};


/**
 * @param {GeoPackageWorkerMessage} originalMsg
 * @param {*=} opt_data
 */
var success = function(originalMsg, opt_data) {
  var msg = {
    message: originalMsg,
    type: MsgType.SUCCESS
  };

  var transferables;

  if (opt_data != null) {
    msg.data = opt_data;

    if (!isNode) {
      if (msg.data instanceof ArrayBuffer) {
        transferables = [msg.data];
      } else if (ArrayBuffer.isView(msg.data)) {
        msg.data = msg.data.buffer;
        transferables = [msg.data];
      }
    }
  }

  postMessage(msg, transferables);
};


/**
 * Map of Geopackages to their IDs.
 * @type {Object<string, Geopackage>}
 */
var gpkgById = {};


/**
 * Map of tableNames to TileScaling objects.
 * @type {Object<string, Geopackage>}
 */
var tileScalingByTableName = {};


/**
 * @param {GeoPackageWorkerMessage} msg
 */
var openGpkg = function(msg) {
  if (!msg.url && !msg.data) {
    handleError('url or data property must exist', msg);
    return;
  }

  if (!msg.id) {
    handleError('id property must exist', msg);
    return;
  }

  if (msg.data) {
    var data = msg.data;
    if (data instanceof ArrayBuffer) {
      data = new Uint8Array(data);
    }

    if (!(data instanceof Uint8Array)) {
      handleError('data must be ArrayBuffer or Uint8Array', msg);
      return;
    }
  } else if (msg.url) {
    if (msg.url.startsWith('file://')) {
      data = decodeURIComponent(msg.url.substring(isWin() ? 8 : 7));
    }
  }

  if (!data) {
    handleError('data or url property must exist', msg);
    return;
  }

  GeoPackage.GeoPackageAPI.open(data)
      .then(function(gpkg) {
        gpkgById[msg.id] = gpkg;
        success(msg);
      })
      .catch(function(err) {
        handleError(err, msg);
      });
};


/**
 * @param {GeoPackageWorkerMessage} msg
 * @return {Geopackage|undefined}
 */
var getGpkg = function(msg) {
  if (!msg.id) {
    handleError('id property must be set', msg);
    return;
  }

  var id = msg.id;
  if (!(id in gpkgById)) {
    handleError('No open GeoPackage exists for the given ID', msg);
    return;
  }

  return gpkgById[id];
};


/**
 * @param {GeoPackageWorkerMessage} msg
 */
var closeGpkg = function(msg) {
  // we get it this way so it does not error if it does not exist
  var gpkg = gpkgById[msg.id];

  if (gpkg) {
    gpkg.close();
    delete gpkgById[msg.id];
  }
};


/**
 * @param {Object} info
 * @return {function(?GeoPackage.TileMatrix):?number}
 */
var getTileMatrixToResolutionMapper = function(info) {
  return (
    /**
     * @param {?GeoPackage.TileMatrix} tileMatrix
     * @return {?number} resolution
     */
    function(tileMatrix) {
      if (tileMatrix) {
        if (tileMatrix.pixel_x_size) {
          return tileMatrix.pixel_x_size;
        } else {
          // compute the pixel_x_size from other values
          return (info.tileMatrixSet.maxX - info.tileMatrixSet.minX) /
              (tileMatrix.matrix_width * tileMatrix.tile_width);
        }
      }

      return null;
    });
};


/**
 * @param {?GeoPackage.TileMatrix} tileMatrix
 * @return {?(number|ol.Size)} The tile size
 */
var tileMatrixToTileSize = function(tileMatrix) {
  if (!tileMatrix) {
    return null;
  }

  var h = tileMatrix.tile_height;
  var w = tileMatrix.tile_width;
  return w === h ? w : [w, h];
};


/**
 * Cesium must have a full tile pyramid (ugh), and so we let it have one and then
 * feed it blank tiles. Due to the full pyramid, we can't have empty sizes on the
 * front of the tile array. Since these are just gonna result in blanks, just use
 * the same as the first defined value.
 * @param {Array<?(number|ol.Size)>} sizes
 * @return {Array<!(number|ol.Size)>} sizes
 */
var fixSizes = function(sizes) {
  var first;
  for (var i = 0, n = sizes.length; i < n; i++) {
    if (sizes[i]) {
      first = sizes[i];
      break;
    }
  }

  while (i--) {
    sizes[i] = first;
  }

  return sizes;
};


/**
 * @param {GeoPackageWorkerMessage} msg
 */
var listDescriptors = function(msg) {
  var gpkg = getGpkg(msg);

  if (gpkg) {
    try {
      var tileTables = gpkg.getTileTables();
      var tileScaling;

      var tilePromises = tileTables.map(async function(tableName) {
        var tileDao = gpkg.getTileDao(tableName);
        var info = gpkg.getInfoForTable(tileDao);

        if (info) {
          var tileMatrices = tileDao.zoomLevelToTileMatrix;
          var config = {
            type: 'geopackage-tile',
            title: info.tableName,
            tableName: info.tableName,
            gpkgMinZoom: Math.round(info.minZoom),
            gpkgMaxZoom: Math.round(info.maxZoom),
            resolutions: tileMatrices.map(getTileMatrixToResolutionMapper(info)),
            tileSizes: fixSizes(tileMatrices.map(tileMatrixToTileSize))
          };

          // Create and store a TileScaling for each tile layer that allows requesting tiles at +4 to -25 zoom levels.
          // This allows tile to show at all zoom levels above where the file defines, but only up to +4 zoom levels
          // higher. Tiles tend to become too blurred once upscaled much more than that.
          tileScaling = new GeoPackage.TileScaling();
          /* eslint-disable google-camelcase/google-camelcase */
          tileScaling.scaling_type = GeoPackage.TileScalingType.IN_OUT;
          tileScaling.zoom_in = 25;
          tileScaling.zoom_out = 4;
          /* eslint-enable google-camelcase/google-camelcase */

          const tileScalingExtension = gpkg.getTileScalingExtension(tableName);
          await tileScalingExtension.getOrCreateExtension();
          tileScalingExtension.createOrUpdate(tileScaling);
          tileScalingByTableName[tableName] = tileScaling;

          if (info.contents) {
            config.title = info.contents.identifier || config.title;
            config.description = info.contents.description || config.description;
          }

          if (info.srs) {
            config.projection = info.srs.organization.toUpperCase() + ':' +
                (info.srs.organization_coordsys_id || info.srs.id);
          }

          if (info.tileMatrixSet) {
            config.extent = [
              info.tileMatrixSet.minX,
              info.tileMatrixSet.minY,
              info.tileMatrixSet.maxX,
              info.tileMatrixSet.maxY];

            config.extentProjection = config.projection || 'EPSG:' + info.tileMatrixSet.srsId;
          }

          return config;
        }
      });

      var featureConfigs = gpkg.getFeatureTables().map(function(tableName) {
        var featureDao = gpkg.getFeatureDao(tableName);
        var info = gpkg.getInfoForTable(featureDao);

        if (info) {
          var cols = info.columns.map(function(col) {
            var dataType = GeoPackage.GeoPackageDataType[col.dataType] || '';
            return /** @type {os.ogc.FeatureTypeColumn} */ ({
              type: dataType.toLowerCase(),
              // iterateGeoJSONFeatures uses the displayName for the feature properties. if the gpkg has a
              // gpkg_data_columns table, col.displayName may differ from col.name. if that table doesn't exist, these
              // fields should be the same. the fallback to col.name is a sanity check, but shouldn't be needed in
              // practice.
              name: col.displayName || col.name
            });
          });

          var config = {
            type: 'geopackage-vector',
            title: info.tableName,
            tableName: info.tableName,
            dbColumns: cols
          };

          if (info.contents) {
            config.title = info.contents.identifier || config.title;
            config.description = info.contents.description || config.description;
          }

          return config;
        }
      });

      // once all the tile layer promises resolve, then call the success callback with the full set of configs
      Promise.all(tilePromises).then((tileConfigs) => {
        success(msg, tileConfigs.concat(featureConfigs));
      }, (error) => {
        handleError(error, msg);
      });
    } catch (e) {
      handleError(e, msg);
      return;
    }
  }
};


/**
 * @param {GeoPackageWorkerMessage} msg
 */
var getTile = function(msg) {
  var gpkg = getGpkg(msg);

  if (!msg.tableName) {
    handleError('tableName property must be set', msg);
    return;
  }

  if (msg.zoom == null) {
    handleError('zoom property must be set', msg);
    return;
  }

  if (!msg.projection) {
    handleError('projection property must be set', msg);
    return;
  }

  if (!msg.width) {
    handleError('width property must be set', msg);
    return;
  }

  if (!msg.height) {
    handleError('height property must be set', msg);
    return;
  }

  if (!msg.extent) {
    handleError('extent (ol.Extent in EPSG:4326) property must be set', msg);
    return;
  }

  try {
    var tileDao = gpkg.getTileDao(msg.tableName);
    var ret = new GeoPackage.GeoPackageTileRetriever(tileDao, msg.width, msg.height);
    var extent = msg.extent.slice();

    // the Geopackage lib doesn't like lat/lons outside of [-180, 180], so normalize the longitude values
    // before constructing our bounding box for each tile
    normalizeTileExtent(extent);

    var bbox = new GeoPackage.BoundingBox(extent[0], extent[2], extent[1], extent[3]);

    // set the scaling for the layer on the retriever to ensure tiles come back outside the default range
    var tileScaling = tileScalingByTableName[msg.tableName];
    if (tileScaling) {
      ret.setScaling(tileScaling);
    }

    ret.getTileWithWgs84BoundsInProjection(bbox, msg.zoom, msg.projection)
        .then(function(tile) {
          success(msg, tile);
        })
        .catch(function(err) {
          handleError(err, msg);
        });
  } catch (e) {
    handleError(e, msg);
  }
};


/**
 * @param {GeoPackageWorkerMessage} msg
 */
var getFeatures = function(msg) {
  var gpkg = getGpkg(msg);

  if (!msg.tableName) {
    handleError('tableName property must be set', msg);
    return;
  }

  try {
    var itr = gpkg.iterateGeoJSONFeatures(msg.tableName);
    var record = itr.next();
    while (record) {
      success(msg, record.value);
      record = record.done ? null : itr.next();
    }

    success(msg, 0);
  } catch (e) {
    handleError(e, msg);
  }
};


/**
 * @param {GeoPackageWorkerMessage} msg
 */
var exportCreate = function(msg) {
  if (!msg.id) {
    handleError('id property must be set', msg);
    return;
  }

  var url = msg.url || (isNode ? 'tmp.gpkg' : undefined);

  GeoPackage.GeoPackageAPI.create(url)
      .then(function(gpkg) {
        if (gpkg) {
          gpkgById[msg.id] = gpkg;
          success(msg);
        }
      })
      .catch(function(err) {
        handleError(err, msg);
      });
};


/**
 * @param {GeoPackageWorkerMessage} msg
 */
var exportCreateTable = function(msg) {
  var gpkg = getGpkg(msg);

  if (!msg.tableName) {
    handleError('tableName property must be set', msg);
    return;
  }

  if (!msg.columns) {
    handleError('columns property must be set', msg);
    return;
  }

  var FeatureColumn = GeoPackage.FeatureColumn;
  var DataType = GeoPackage.GeoPackageDataType;
  var GeometryType = GeoPackage.GeometryType;

  // disable camelcase checks for external library that does not conform to the rule
  /* eslint-disable google-camelcase/google-camelcase */
  var geometryColumns = new GeoPackage.GeometryColumns();
  geometryColumns.table_name = msg.tableName;
  geometryColumns.column_name = 'geometry';
  geometryColumns.geometry_type_name = GeometryType.nameFromType(GeometryType.GEOMETRY);
  geometryColumns.z = 2;
  geometryColumns.m = 0;
  /* eslint-enable google-camelcase/google-camelcase */

  var columns = [];
  columns.push(FeatureColumn.createPrimaryKeyColumn(0, 'id'));
  columns.push(FeatureColumn.createGeometryColumn(1, 'geometry', GeometryType.GEOMETRY, false, null));

  msg.columns.forEach(function(col) {
    if (col.field.toLowerCase() === 'id' || col.field.toLowerCase() === 'geometry' ||
        col.field === 'TIME_START' || col.field === 'TIME_STOP') {
      return;
    }

    if (col.field === 'recordTime') {
      columns.push(FeatureColumn.createColumn(columns.length, 'TIME_START', DataType.DATETIME, false, null));
      columns.push(FeatureColumn.createColumn(columns.length, 'TIME_STOP', DataType.DATETIME, false, null));
    } else {
      var type = DataType.TEXT;
      var defaultValue = '';
      var colType = col.type.toLowerCase();

      if (colType === 'decimal') {
        type = DataType.REAL;
        defaultValue = null;
      } else if (colType === 'integer') {
        type = DataType.INTEGER;
        defaultValue = null;
      } else if (colType === 'datetime') {
        type = DataType.TEXT;
        defaultValue = null;
      }

      columns.push(FeatureColumn.createColumn(columns.length, col.field, type, false, defaultValue));
    }
  });

  try {
    gpkg.createFeatureTable(msg.tableName, geometryColumns, columns);
    success(msg);
  } catch (e) {
    handleError(e, msg);
  }
};


/**
 * @param {GeoPackageWorkerMessage} msg
 */
var exportGeoJSON = function(msg) {
  var gpkg = getGpkg(msg);

  if (!msg.tableName) {
    handleError('tableName property must be set', msg);
    return;
  }

  if (!msg.data || typeof msg.data !== 'object') {
    handleError('GeoJSON feature not found on msg.data', msg);
    return;
  }

  try {
    var geojson = msg.data;
    var features = geojson.features;

    features.forEach((feature) => {
      var props = feature.properties;

      // time start and stop are ISO8601 strings, and the new API needs dates
      if ('TIME_START' in props) {
        props.TIME_START = new Date(Date.parse(props.TIME_START));
      }

      if ('TIME_STOP' in props) {
        props.TIME_STOP = new Date(Date.parse(props.TIME_STOP));
      }
    });

    /**
     * Progress callback function for updating the UI with the current parse count.
     * @param {number} count Current parse count.
     */
    var progressFn = (count) => {
      // post a progress message with the current count
      success(/** @type {GeoPackageWorkerMessage} */ ({
        id: msg.id,
        type: msg.type,
        command: 'progress',
        tableName: msg.tableName,
        count
      }));
    };

    gpkg.addGeoJSONFeaturesToGeoPackage(features, msg.tableName, false, 10000, progressFn).then((numAdded) => {
      success(msg);
    }, (reason) => {
      handleError(reason, msg);
    });
  } catch (e) {
    handleError(e, msg);
  }
};


/**
 * @type {Object<string, {data: Uint8Array, index: number}>}
 */
var exportsById = {};


/**
 * @param {GeoPackageWorkerMessage} msg
 */
var exportWrite = function(msg) {
  var gpkg = getGpkg(msg);

  gpkg.export().then((data) => {
    exportsById[msg.id] = {
      data: new Uint8Array(data),
      index: 0
    };

    success(msg);
  }, (err) => {
    handleError(err, msg);
  });
};


/**
 * @param {GeoPackageWorkerMessage} msg
 */
var exportGetChunk = function(msg) {
  if (!msg.id) {
    handleError('id property must be set', msg);
    return;
  }

  var ex = exportsById[msg.id];

  if (!ex) {
    handleError('an export for the id has not been started', msg);
    return;
  }

  var data = ex.data;

  if (isNode) {
    var limit = Math.min(ex.index + (1024 * 1024) + 1, ex.data.length);
    var data = Array.from(ex.data.subarray(ex.index, limit));
    ex.index = limit;
  }

  success(msg, data);
};


/**
 * @param {GeoPackageWorkerMessage} msg
 */
var exportWriteFinish = function(msg) {
  closeGpkg(msg);
  delete exportsById[msg.id];
  success(msg);
};


var ExportCommands = {
  create: exportCreate,
  createTable: exportCreateTable,
  geojson: exportGeoJSON,
  write: exportWrite,
  getChunk: exportGetChunk,
  writeFinish: exportWriteFinish
};


/**
 * @param {GeoPackageWorkerMessage} msg
 */
var exportGpkg = function(msg) {
  if (!msg.command) {
    handleError('command property must be set', msg);
    return;
  }

  if (msg.command in ExportCommands) {
    ExportCommands[msg.command](msg);
  } else {
    handleError('Unknown command type', msg);
  }
};


/**
 * @param {GeoPackageWorkerMessage} msg
 */
var openLibrary = function(msg) {
  if (!isNode) {
    // this allows the main application to detect where this is loaded
    importScripts(msg.url);
    GeoPackage = self.GeoPackage;
    setupLibrary(msg.url.replace('/geopackage.min.js', ''));
  }
};


/**
 * If the OffscreenCanvas API is supported.
 * @return {boolean}
 */
var supportsOffscreenCanvas = function() {
  if (typeof self === 'object' && self.OffscreenCanvas) {
    try {
      var canvas = new OffscreenCanvas(0, 0);
      var context = canvas.getContext('2d');
      return !!context;
    } catch (e) {}
  }

  return false;
};


/**
 * Set up the geopackage library
 * @param {string} distPath Path to the dist folder.
 * @param {string=} pathSep The path separator. Defaults to '/'.
 */
var setupLibrary = function(distPath, pathSep = '/') {
  GeoPackage.setSqljsWasmLocateFile((file) => [distPath, file].join(pathSep));
  GeoPackage.setCanvasKitWasmLocateFile((file) => [distPath, 'canvaskit', file].join(pathSep));

  // Prefer the OffscreenCanvas browser API if available, to avoid loading the CanvasKit library.
  var canvasAdapter = supportsOffscreenCanvas() ? GeoPackage.OffscreenCanvasAdapter : GeoPackage.CanvasKitCanvasAdapter;
  GeoPackage.Context.setupCustomContext(GeoPackage.SqljsAdapter, canvasAdapter);
};


var MsgCommands = {
  openLibrary: openLibrary,
  open: openGpkg,
  close: closeGpkg,
  listDescriptors: listDescriptors,
  getTile: getTile,
  getFeatures: getFeatures,
  export: exportGpkg
};


/**
 * @param {Event|GeoPackageWorkerMessage} evt The message
 * @this Worker
 */
var onMessage = function(evt) {
  var msg = /** @type {GeoPackageWorkerMessage} */ (isNode ? evt : evt.data);

  if (msg) {
    if (msg.type in MsgCommands) {
      MsgCommands[msg.type](msg);
    } else {
      handleError('Unknown message type', msg);
    }
  }
};


(function() {
  if (typeof self === 'object') {
    self.addEventListener('message', onMessage);
  } else {
    isNode = true;
    process.on('message', onMessage);

    /**
     * @param {GeoPackageWorkerResponse} msg
     */
    global.postMessage = function(msg) {
      process.send(msg);
    };

    // CLEVER HACK ALERT!
    // This script runs in either a Worker (web) or in a node child process (Electron).
    // The child process has a node-only environment by default, rather than an Electron
    // environment. However, electron-builder only packages the version built for the
    // Electron environment.
    //
    // Therefore, trick node-pre-gyp into thinking we're in Electron.
    // see associated env variable set in geopackage.js
    if (process.versions.electron == null && process.env.ELECTRON_VERSION) {
      process.versions.electron = process.env.ELECTRON_VERSION;
    }

    if (process.env.ELECTRON_EXTRA_PATH) {
      module.paths.unshift(process.env.ELECTRON_EXTRA_PATH);
    }

    GeoPackage = require('@ngageoint/geopackage');

    var path = require('path');
    var basePath = require.resolve('@ngageoint/geopackage');
    var distPath = path.dirname(basePath);
    setupLibrary(distPath, path.sep);
  }
})();
